helm: allow annotations to be set for preflight resources

bradwhitfield opened the pull request:
Currently preflight resources, except the serviceaccount, do not allow annotations to be configured at the top level of the resources. This patch adds support for adding annotations.

Adding annotations at the top level of resources (not just in pod templates) will allow for use cases like ArgoCD and helm hooks.

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

```release-note
helm: allow annotations to be set for preflight resources
```

bradwhitfield committed:
helm: allow annotations to be set for preflight resources

Currently preflight resources, except the serviceaccount, do not allow annotations to be configured at the top level of the resources. This patch adds support for adding annotations.

Adding annotations at the top level of resources (not just in pod templates) will allow for use cases like ArgoCD and helm hooks.

Signed-off-by: Brad Whitfield <bradswhitfield@gmail.com>

gandro reviewed (approved):
Thanks! PR and motivation look solid overall. There is an issue with CI (see previous comment) that needs to be fixed, but otherwise this looks good to me

youngnick reviewed (approved):
PR LGTM once the CI is sorted, nice work @bradwhitfield.

bradwhitfield committed:
helm: allow annotations to be set on nearly all resources

Currently most resources don't allow for setting custom annotations at the top level of the resources. This change allows chart consumers to define annotations for these resources breaking them up by resources under "templates" and additional fine-grained annotations for `tls.auto` resources.

Adding annotations at the top level of resources (not just in pod templates) will allow for use cases like ArgoCD and helm hooks.

Signed-off-by: Brad Whitfield <bradswhitfield@gmail.com>

gandro commented on Documentation/helm-values.rst:
```diff
@@ -2472,6 +2472,10 @@
      - Affinity for cilium-preflight
      - object
      - ``{"podAffinity":{"requiredDuringSchedulingIgnoredDuringExecution":[{"labelSelector":{"matchLabels":{"k8s-app":"cilium"}},"topologyKey":"kubernetes.io/hostname"}]}}``
+   * - :spelling:ignore:`preflight.annotations`
```
nit: Please squash this into the previous commit

gandro commented on install/kubernetes/cilium/README.md:
```diff
@@ -60,12 +60,14 @@ contributors across the globe, there is almost always someone available to help.
 | aksbyocni.enabled | bool | `false` | Enable AKS BYOCNI integration. Note that this is incompatible with AKS clusters not created in BYOCNI mode: use Azure integration (`azure.enabled`) instead. |
 | alibabacloud.enabled | bool | `false` | Enable AlibabaCloud ENI integration |
 | annotateK8sNode | bool | `false` | Annotate k8s node upon initialization with Cilium's metadata. |
+| annotations | object | `{}` | Annotations to be added to all top-level cilium-agent objects (resources under templates/cilium-agent) |
```
Same as above: Please squash into the previous commit. That way we can use "git bisect" to build without errors

bradwhitfield replied:
So to make sure I'm clear, should I end up with two commits like these, or just one commit like the bottom one?
* helm: allow annotations to be set for preflight resources 
* helm: allow annotations to be set on nearly all resources

edit:

For the PR feedback, should that be in another commit, or would you like it amended to the `helm: allow annotations to be set on nearly all resources` commit?

gandro replied:
I think just having the two commits that you outlined will yield a clean git history, that sounds good to me. No need to put the PR feedback in a separate commit, since GitHub can show us those diffs too (as long as you don't change the commits and rebase at the same time).

gandro commented on install/kubernetes/cilium/templates/cilium-envoy/service.yaml:
```diff
@@ -4,10 +4,16 @@ kind: Service
   name: cilium-envoy
   namespace: {{ .Release.Namespace }}
-  {{- if not .Values.envoy.prometheus.serviceMonitor.enabled }}
+  # TODO CONFIRM THIS SYNTAX
```
This looks correct to me

bradwhitfield replied:
Didn't meant to leave that in. I'll clean that up. Good catch.

gandro commented on install/kubernetes/cilium/templates/cilium-nodeinit/serviceaccount.yaml:
```diff
@@ -4,8 +4,13 @@ kind: ServiceAccount
   name: {{ .Values.serviceAccounts.nodeinit.name | quote }}
   namespace: {{ .Release.Namespace }}
-  {{- if .Values.serviceAccounts.nodeinit.annotations }}
+  {{- if or .Values.serviceAccounts.nodeinit.annotations .Values.annotations }}
```
This should be `.Values.nodeinit.annotations` in the second expression

gandro commented on install/kubernetes/cilium/templates/clustermesh-apiserver/tls-certmanager/admin-secret.yaml:
```diff
@@ -5,6 +5,10 @@ kind: Certificate
 metadata:
   name: clustermesh-apiserver-admin-cert
   namespace: {{ .Release.Namespace }}
+  {{- with .Values.clustermesh.apiserver.tls.auto.annotations }}
```
I think we probably can just use the overall clustermesh-api-server annotations here, i.e. `.Values.clustermesh.annotations`, same for all other clustermesh TLS resources

gandro commented on install/kubernetes/cilium/templates/hubble/tls-certmanager/relay-client-secret.yaml:
```diff
@@ -5,6 +5,10 @@ kind: Certificate
 metadata:
   name: hubble-relay-client-certs
   namespace: {{ .Release.Namespace }}
+  {{- with .Values.hubble.tls.auto.annotations }}
```
I think we could just use `.Values.hubble.annotations` for all TLS related objects (i.e. templates/hubble/tls-*). It would make it a bit easier I think

bradwhitfield committed:
helm: simplify auto TLS annotations and various cleanup based on PR feedback

Signed-off-by: Brad Whitfield <bradswhitfield@gmail.com>

gandro reviewed (approved):
Code-wise this looks good to me now, thanks!

aditighag reviewed (approved):
Approving for my codeowners

joestringer commented on install/kubernetes/Makefile:
```diff
@@ -44,7 +44,7 @@ update-chart: $(ROOT_DIR)/VERSION # Update chart versions to point to the curren
 	$(QUIET)grep -lR -e 'version:' -e 'appVersion:' $(CILIUM_CHARTS)/ \
 		| xargs -L 1 $(SED) -i -e 's/'$(CILIUM_CHART_REGEX)'/\1 '$(VERSION)'/g'
-	$(QUIET)$(SED) -i 's;icon:.*;icon: $(LOGO_BASE_URL)/cilium@$(CILIUM_BRANCH)/$(LOGO_PATH);' $(CHART_FILE)
+	# $(QUIET)$(SED) -i 's;icon:.*;icon: $(LOGO_BASE_URL)/cilium@$(CILIUM_BRANCH)/$(LOGO_PATH);' $(CHART_FILE)
```
Why was this change made? It doesn't seem to be related to the goal of the PR.

joestringer replied:
@gandro @youngnick @aditighag I know this is a nit, but how is this falling through review?

aditighag replied:
It was probably an oversight. Based on my update, I only looked at the files for my CODEOWNERS (contributing), so I most likely didn't look at the changes in this file. 

`/install/kubernetes/ @cilium/sig-k8s @cilium/helm`

bradwhitfield replied:
That was not intentional. I had to fight a bit to get the makefiles to work on my computer. I eventually did. I don't remember this change specifically, but I probably commented it out for testing and didn't catch that I committed it. I can open a PR to fix that if you'd like.

gandro replied:
@bradwhitfield Yeah, a PR to fix it would be appreciated.
Also https://github.com/cilium/cilium/pull/27860#discussion_r1378867381 sounds like something we should address in a follow-up PR.

@joestringer I'm not sure how this slipped during review. It seems the [commit which did the change in question here](https://github.com/cilium/cilium/pull/27860/commits/dea1d189e46945f5f0cfffeae366f3e103ba9999) was pushed after I already acked the PR. I should have performed a re-review before running CI, I assumed the changes were mostly due to rebases. Sorry about that.

bradwhitfield replied:
I submitted #28929. The checks are passing and I believe I did the "fixes" correctly.

joestringer replied:
Thanks all :+1: My goal with starting the thread was to understand whether there's some underlying issue on some platforms that might have motivated this change in the first place (in which case maybe we can tweak the makefiles/scripts to be more platform-friendly?) and to get a better understanding of why unintentional changes may be introduced into the tree.

While in general I think we have a policy not to add commented code to the tree, it is ultimately up to the reviewers to enforce these constraints and I'd typically expect reviewers to collectively lay eyes on every line of code and have an understanding about how that code fits into the broader picture. In this case it seems like the submission slipped through both Brad's filter as he submitted the PR and the reviewer's filters as they accepted the PR. 

There's always a small chance of this sort of thing happening, and I understand that contributors can make mistakes or have a slip of oversight, particularly when changes are introduced later in the review cycle. Given the brief retrospective we've collectively done here, I don't think there is a systematic issue to be concerned about beyond this PR; it's just a classic case of not realizing that there was another change that came in after initial review which we would usually not accept. Fortunately the consequences in this case are low.

I hope that this discussion has helped with sharing understandings and expectations around the PR submission process going forward. I appreciate the quick feedback & solution @bradwhitfield :pray:

coro commented on install/kubernetes/cilium/templates/cilium-agent/serviceaccount.yaml:
```diff
@@ -8,4 +8,13 @@ metadata:
   annotations:
     {{- toYaml .Values.serviceAccounts.cilium.annotations | nindent 4 }}
   {{- end }}
+  {{- if or .Values.serviceAccounts.cilium.annotations .Values.annotations }}
+  annotations:
+    {{- with .Values.annotations }}
+      {{- toYaml . | nindent 4 }}
+    {{- end }}
+    {{- with .Values.serviceAccounts.cilium.annotations }}
+      {{- toYaml . | nindent 4 }}
+    {{- end }}
+  {{- end }}
```
@bradwhitfield This is causing a duplicate `annotations` key if `.Values.serviceAccounts.cilium.annotations` is specified in the Helm release. Would you mind sending in a PR to fix this?
